File: pele/potentials/fortran/lj.f90
subroutine ljenergy( coords, natoms, e, eps, sig, periodic, boxl )
implicit none
integer, intent(in) :: natoms
double precision, intent(in) :: coords(3*natoms), sig, eps, boxl
double precision, intent(out) :: e
double precision dr(3), sig6, sig12, r2, ir2, ir6, ir12, iboxl
logical, intent(in) :: periodic
integer j1, j2

if (periodic) iboxl = 1.d0 / boxl

sig6 = sig**6
sig12 = sig6*sig6

e = 0.d0
do j1 = 1,natoms
   do j2 = 1,j1-1
      dr(:) = coords(3*(j1-1)+1 : 3*(j1-1) + 3) - coords(3*(j2-1)+1 : 3*(j2-1) + 3)
      if (periodic)  dr(:) = dr(:) - nint( dr(:) * iboxl ) * boxl
      r2 = sum( dr(:)**2 )
      ir2 = 1.d0/r2
      ir6 = ir2**3
      ir12 = ir6**2
      e = e - 4.d0 * eps * (sig6*ir6 - sig12*ir12)
   enddo
enddo

end subroutine ljenergy

subroutine ljenergy_gradient( coords, natoms, e, grad, eps, sig, periodic, boxl )
implicit none
integer, intent(in) :: natoms
double precision, intent(in) :: coords(3*natoms), sig, eps, boxl
double precision, intent(out) :: e, grad(3*natoms)
logical, intent(in) :: periodic
double precision dr(3), sig6, sig12, r2, ir2, ir6, ir12, g, iboxl
integer j1, j2, i1, i2

if (periodic) iboxl = 1.d0 / boxl

sig6 = sig**6
sig12 = sig6*sig6

e = 0.d0
grad(:) = 0.d0
do j1 = 1,natoms
   i1 = 3*(j1-1)
   do j2 = 1,j1-1
      i2 = 3*(j2-1)
      dr(:) = coords(i1+1 : i1 + 3) - coords(i2+1 : i2 + 3)
      if (periodic)  dr(:) = dr(:) - nint( dr(:) * iboxl ) * boxl
      r2 = sum( dr(:)**2 )
      ir2 = 1.d0/r2
      ir6 = ir2**3
      ir12 = ir6**2
      e = e - 4.d0 * eps * (sig6*ir6 - sig12*ir12)

      g = 4.d0 * eps * (12.d0 * sig12 * ir12 -  6.d0 * sig6 * ir6) * ir2;
      grad(i1+1 : i1+3) = grad(i1+1 : i1+3) - g * dr(:)
      grad(i2+1 : i2+3) = grad(i2+1 : i2+3) + g * dr(:)
   enddo
enddo
end subroutine ljenergy_gradient

subroutine energy_ilist( coords, natoms, e, eps, sig, ilist, nlist, periodic, boxl )
implicit none
integer, intent(in) :: natoms, nlist, ilist(nlist* 2)
double precision, intent(in) :: coords(3*natoms), sig, eps, boxl
double precision, intent(out) :: e
logical, intent(in) :: periodic
double precision dr(3), sig6, sig12, r2, ir2, ir6, ir12, iboxl
integer j1, j2, i1, i2, n

if (periodic) iboxl = 1.d0 / boxl

sig6 = sig**6
sig12 = sig6*sig6

e = 0.d0

do n = 1,nlist
   j1 = ilist(2*(n-1)+1) + 1 !convert to fortran indexing
   j2 = ilist(2*(n-1)+2) + 1 !convert to fortran indexing

   i1 = 3*(j1-1)
   i2 = 3*(j2-1)

   dr(:) = coords(i1+1 : i1 + 3) - coords(i2+1 : i2 + 3)
   if (periodic)  dr(:) = dr(:) - nint( dr(:) * iboxl ) * boxl
   r2 = sum( dr(:)**2 )
   ir2 = 1.d0/r2
   ir6 = ir2**3
   ir12 = ir6**2
   e = e - 4.d0 * eps * (sig6*ir6 - sig12*ir12)

enddo
end subroutine energy_ilist

subroutine energy_gradient_ilist( coords, natoms, e, grad, eps, sig, ilist, nlist, periodic, boxl )
implicit none
integer, intent(in) :: natoms, nlist, ilist(nlist* 2)
double precision, intent(in) :: coords(3*natoms), sig, eps, boxl
double precision, intent(out) :: e, grad(3*natoms)
logical, intent(in) :: periodic
double precision dr(3), sig6, sig12, r2, ir2, ir6, ir12, g, iboxl
integer j1, j2, i1, i2, n

if (periodic) iboxl = 1.d0 / boxl

sig6 = sig**6
sig12 = sig6*sig6

e = 0.d0
grad(:) = 0.d0

do n = 1,nlist
   j1 = ilist(2*(n-1)+1) + 1 !convert to fortran indexing
   j2 = ilist(2*(n-1)+2) + 1 !convert to fortran indexing

   i1 = 3*(j1-1)
   i2 = 3*(j2-1)

   dr(:) = coords(i1+1 : i1 + 3) - coords(i2+1 : i2 + 3)
   if (periodic)  dr(:) = dr(:) - nint( dr(:) * iboxl ) * boxl
   r2 = sum( dr(:)**2 )
   ir2 = 1.d0/r2
   ir6 = ir2**3
   ir12 = ir6**2
   e = e - 4.d0 * eps * (sig6*ir6 - sig12*ir12)

   g = 4.d0 * eps * (12.d0 * sig12 * ir12 -  6.d0 * sig6 * ir6) * ir2;
   grad(i1+1 : i1+3) = grad(i1+1 : i1+3) - g * dr(:)
   grad(i2+1 : i2+3) = grad(i2+1 : i2+3) + g * dr(:)
enddo
end subroutine energy_gradient_ilist

File: pele/potentials/lj.py
from __future__ import print_function
from __future__ import absolute_import
import numpy as np

from pele.potentials import BasePotential
from .fortran import lj as ljf


__all__ = ["LJ"]


class LJ(BasePotential):
    """ simple lennard jones potential"""

    def __init__(self, eps=1.0, sig=1.0, boxl=None):
        self.sig = sig
        self.eps = eps
        self.boxl = boxl
        if self.boxl is None:
            self.periodic = False
            self.boxl = 10000.
        else:
            self.periodic = True

    def getEnergy(self, coords):
        E = ljf.ljenergy(
            coords, self.eps, self.sig, self.periodic, self.boxl)
        return E

    def getEnergyGradient(self, coords):
        E, grad = ljf.ljenergy_gradient(
            coords, self.eps, self.sig, self.periodic, self.boxl)
        return E, grad

    def getEnergyList(self, coords, ilist):
        E = ljf.energy_ilist(
            coords, self.eps, self.sig, ilist.reshape(-1), self.periodic,
            self.boxl)
        return E

    def getEnergyGradientList(self, coords, ilist):
        E, grad = ljf.energy_gradient_ilist(
            coords, self.eps, self.sig, ilist.reshape(-1), self.periodic,
            self.boxl)
        return E, grad

    def getEnergyGradientHessian(self, coords):
        if self.periodic: raise Exception("Hessian not implemented for periodic boundaries")
        from .fortran.lj_hess import ljdiff

        g, energy, hess = ljdiff(coords, True, True)
        return energy, g, hess


#
# only testing below here
#

def main():  # pragma: no cover
    # test class
    natoms = 12
    coords = np.random.uniform(-1, 1, natoms * 3) * 2

    lj = LJ()
    E = lj.getEnergy(coords)
    print("E", E)
    E, V = lj.getEnergyGradient(coords)
    print("E", E)
    print("V")
    print(V)

    print("try a quench")
    from pele.optimize import mylbfgs as quench

    quench(coords, lj, iprint=1)
    #quench( coords, lj.getEnergyGradientNumerical, iprint=1 )


if __name__ == "__main__":
    main()

